### ISSUE TITLE
Example of using Arrow C interface

### ISSUE THREAD

**Issue by kesavkolla** [wangfenjin/duckdb-rs#79]
Can someone provide an example of how to seemlessly pass the arrow to duckdb in rust? I have the location of parquet file(s) in Azure ADLSv2. I am currently using object_store and arrow-rs to read those parquet files. I do see duckdb and arrow work very seemless in Python. Is there some example that someone can point how to use in rust.

Thanks in advance.

**Comment by wangfenjin** [comment]
you mean something like this? https://duckdb.org/docs/guides/python/sql_on_arrow

Better share your python examples

**Comment by kesavkolla** [comment]
Yes that is correct. The example given in that URL. Ideally the recordbatchreader querying on it would be super useful from rust.

**Comment by wangfenjin** [comment]
Haven't support that in rust api. For now what you can do is load parquet directly into duckdb

#80

**Comment by kesavkolla** [comment]
Hmmm... unfortunately I can't load parquet in duckdb because duckdb doesn't have support for Azure ADLS. Currently duckdb has no pluggable mechanism for loading parquet files.

Is it possible to write an extension for duckdb in rust?

**Comment by wangfenjin** [comment]
What about download parquet first and then load it into duckdb?

I haven't spent too much time to dive deep into the duckdb extension, need time to investigate

**Comment by kesavkolla** [comment]
Yeah I could do that. Ideally it would be nice to provide Arrow C interface so we can easily integrate between rust and duckdb

**Comment by wangfenjin** [comment]
Yes, the API would be much simpler. 

But the logic still the same, download the file, and load into duckdb. Read into arrow and then bind the data into duckdb might be slower than directly read parquet file into duckdb

**Comment by kesavkolla** [comment]
I am looking at this example in Python:


```python
from adlfs import AzureBlobFileSystem
import duckdb
import pyarrow.parquet as pq
import pyarrow.dataset as ds

account_name = "xxxxx"
account_key = "xxxxx"
abfs = AzureBlobFileSystem( account_name = account_name, account_key = account_key, container_name = "data")
pqdata = ds.dataset("path/inside/abfs", filesystem=abfs)


conn = duckdb.connect(":memory:")
conn.execute("SELECT * from pqdata LIMIT 10")
```

Something similar in rust would be greatly helpful

**Comment by wangfenjin** [comment]
Acctually this example may not work because it [missed one important steps](https://github.com/duckdb/duckdb/blob/0d13deb759dd993ce037d37d79cc419c725553cd/tools/pythonpkg/tests/fast/arrow/test_dataset.py#L73): 

```python
       conn.register("pqdata",pqdata)
```

And this `register` API is what we need to implement in rust.

And this register API is only useful when you already have some data in memory and want to use duckdb query it. If you only want to load parquet and query using duckdb, using parquet extension is much more straightforward and faster.

```sql
-- load parquest files in data folder
SELECT * FROM read_parquet('./data/*.parquet');
```

**Comment by kesavkolla** [comment]
@wangfenjin Can you provide some high level flow for writing _register_ API. I want to implement that and provide PR

**Comment by wangfenjin** [comment]
hi @kesavkolla , seems we don't have capi yet, I create an issue in duckdb repo, let's wait for the response.

**Comment by snth** [comment]
Hi,

sorry to be chiming in here but I have also been struggling to make sense of these `arrow` interfaces to other systems in Rust so perhaps I can help out a bit. It's an unfortunate that while the Python documentation for most projects is excellent, there are very few docs and examples for Rust.

I've extracted the minimal example from my code (so not 100% sure that it works as I combined a few things but I think it should give a good idea).

```rust
use arrow::record_batch::RecordBatch;
use arrow::util::pretty::pretty_format_batches;
use duckdb::Connection;

let conn = Connection::open_in_memory()?;

let mut stmt = conn.prepare(&sql)?;

let batches = stmt.query_arrow([])?.collect::<Vec<RecordBatch>>();

println!(pretty_format_batches(&batches)?);
```

I think this shows how you can get the data from DuckDB into `arrow` data structures and you can then work with them using other `arrow` functions, e.g. `pretty_format_batches()` in this example.

HTH

**Comment by wangfenjin** [comment]
Yeah I agree that the documentation is not so clear or complete, I may make some time to improve it but I can't guarantee the time. Also I'm not a native English speaker which will spend more time for me to write the doc.

 BTW, if you are looking for examples, you can refer to the test code such as [query_arrow](https://github.com/wangfenjin/duckdb-rs/blob/main/src/lib.rs#L1201) . All public API should have at least one test case, which can also serve as examples.

**Comment by kesavkolla** [comment]
@wangfenjin I looked into the duckdb python code. Here is what python code is doing for registering arrow:

```c
unique_ptr<DuckDBPyRelation> DuckDBPyConnection::FromArrow(py::object &arrow_object) {
	if (!connection) {
		throw ConnectionException("Connection has already been closed");
	}
	py::gil_scoped_acquire acquire;
	string name = "arrow_object_" + GenerateRandomName();
	if (!IsAcceptedArrowObject(arrow_object)) {
		auto py_object_type = string(py::str(arrow_object.get_type().attr("__name__")));
		throw InvalidInputException("Python Object Type %s is not an accepted Arrow Object.", py_object_type);
	}
	auto stream_factory =
	    make_unique<PythonTableArrowArrayStreamFactory>(arrow_object.ptr(), connection->context->config);

	auto stream_factory_produce = PythonTableArrowArrayStreamFactory::Produce;
	auto stream_factory_get_schema = PythonTableArrowArrayStreamFactory::GetSchema;

	auto rel = make_unique<DuckDBPyRelation>(
	    connection
	        ->TableFunction("arrow_scan", {Value::POINTER((uintptr_t)stream_factory.get()),
	                                       Value::POINTER((uintptr_t)stream_factory_produce),
	                                       Value::POINTER((uintptr_t)stream_factory_get_schema)})
	        ->Alias(name));
	rel->rel->extra_dependencies =
	    make_unique<PythonDependencies>(make_unique<RegisteredArrow>(move(stream_factory), arrow_object));
	return rel;
}
```

Also from the ducdb capi test code there is a way to register table function:

https://github.com/duckdb/duckdb/blob/master/test/api/capi/capi_table_functions.cpp#L53

```c
static void capi_register_table_function(duckdb_connection connection, const char *name,
                                         duckdb_table_function_bind_t bind, duckdb_table_function_init_t init,
                                         duckdb_table_function_t f) {
	duckdb_state status;

	// create a table function
	auto function = duckdb_create_table_function();
	duckdb_table_function_set_name(nullptr, name);
	duckdb_table_function_set_name(function, nullptr);
	duckdb_table_function_set_name(function, name);
	duckdb_table_function_set_name(function, name);

	// add a string parameter
	duckdb_logical_type type = duckdb_create_logical_type(DUCKDB_TYPE_BIGINT);
	duckdb_table_function_add_parameter(function, type);
	duckdb_destroy_logical_type(&type);

	// set up the function pointers
	duckdb_table_function_set_bind(function, bind);
	duckdb_table_function_set_init(function, init);
	duckdb_table_function_set_function(function, f);

	// register and cleanup
	status = duckdb_register_table_function(connection, function);
	REQUIRE(status == DuckDBSuccess);

	duckdb_destroy_table_function(&function);
	duckdb_destroy_table_function(&function);
	duckdb_destroy_table_function(nullptr);
}
```

Looks like they do have this register functionality. I do see these **duckdb_register_table_function** in the libduck-sys crate that is part of this repo. 

I am not too familiar on how to use FFI in rust can you add this register function to InnerConnection and Connection? Then we can pass arrow to duckdb.

Thanks in advance

**Comment by duarten** [comment]
Are there plans to export `ArrowVTab`? The code is working fine for me so far.

**Comment by phillipleblanc** [comment]
> Are there plans to export `ArrowVTab`? The code is working fine for me so far.

This is exported now! https://github.com/duckdb/duckdb-rs/pull/259

**Comment by navdeep710** [comment]
@kesavkolla can you help in any updates or hints if you were able to query arrow via duckdb in rust, we are also trying to achieve the same for differnt format not suported by duckdb ?

**Comment by ajwerner** [comment]
Any updates on this?

**Comment by mlafeldt** [comment]
This is now possible and covered by #804.

It adds an `ArrowVTab` example/test for querying Arrow `RecordBatch` data from DuckDB SQL, plus `libduckdb-sys` coverage for the newer Arrow C Data Interface conversion APIs. It also updates the repo to avoid DuckDB's deprecated Arrow result APIs everywhere.
